File: PcodeFiles/Main/src/main/java/pcodefiles/model/Group.java
package pcodefiles.model;

import ghidra.program.model.address.Address;
import ghidra.program.model.address.AddressSpace;
import ghidra.program.model.address.GenericAddress;

import java.util.HashMap;
import java.util.Map;

public class Group {
    protected String id;
    String name;
    Address address;
    enum DataOrg {
        eByte,
        eLoHi, eHiLo,
        eLoHiLoHi, eHiloHilo,
        eFloatLoHi, eFloatHiLo
    }
    DataOrg dataOrg;
    enum GroupType {
        GROUP_TYPE_LIST,
        GROUP_TYPE_MAP_2D,
        GROUP_TYPE_MAP_3D
    }
    GroupType type;
    int x;
    int y;

    public String getFolderName() {
        return folderName;
    }

    public void setFolderName(String folderName) {
        this.folderName = folderName;
    }

    String folderName;

    public Group(String id) {
        this.id = id;
    }

    public Group(String id, String name, Address address, String dataOrg, int type, int length) {
        this.id = id;
        this.name = name;
        this.address = address;
        this.y = length;

        this.type = intToType(type);
        this.dataOrg = stringToDataOrg(dataOrg);
    }

    private GroupType intToType(int type) {
        switch (type) {
            case 0:
                return GroupType.GROUP_TYPE_LIST;
            case 1:
                return GroupType.GROUP_TYPE_MAP_2D;
            case 2:
                return GroupType.GROUP_TYPE_MAP_3D;
        }
        return GroupType.GROUP_TYPE_LIST;
    }

    private DataOrg stringToDataOrg(String dataOrg) {
        switch (dataOrg) {
            case "eByte":
                return DataOrg.eByte;
            case "eLoHi":
                return DataOrg.eLoHi;
            case "eHiLo":
                return DataOrg.eHiLo;
            case "eLoHiLoHi":
                return DataOrg.eLoHiLoHi;
            case "eHiLoHilo":
                return DataOrg.eHiloHilo;
            case "eFloatLoHi":
                return DataOrg.eFloatLoHi;
            case "eFloatHiLo":
                return DataOrg.eFloatHiLo;
        }

        return DataOrg.eByte;
    }

    public String getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public Address getAddress() {
        return address;
    }

    public DataOrg getDataOrg() {
        return dataOrg;
    }

    public GroupType getGroupType() {
        return type;
    }

    public Map<String,Integer> getSizes() {
        var sizes = new HashMap<String, Integer>();
        sizes.put("x", x);
        sizes.put("y", y);

        return sizes;
    }

    public void setSizes(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setAddress(Address address) {
        this.address = address;
    }

    public void setDataOrg(DataOrg dataOrg) {
        this.dataOrg = dataOrg;
    }

    public void setDataOrg(String dataOrg) {
        this.dataOrg = stringToDataOrg(dataOrg);
    }

    public void setGroupType(GroupType type) {
        this.type = type;
    }

    public void setGroupType(int type) {
        this.type = intToType(type);
    }

    public int getDataTypeSize()
    {
        switch (this.dataOrg) {
            case eByte:
                return 1;
            case eHiLo:
            case eLoHi:
                return 2;
            case eLoHiLoHi:
            case eHiloHilo:
                return 4;
            case eFloatHiLo:
            case eFloatLoHi:
                return 8;
        }

        return 1;
    }

    public static Group unserialize(String[] data) {
        return new Group(
                data[0],
                data[1],
                null,
                data[3],
                Integer.parseInt(data[2]),
                Integer.parseInt(data[5])
        );
    }
}

File: PcodeFiles/Main/src/main/java/pcodefiles/model/Group.kt
package pcodefiles.model

import ghidra.program.model.address.Address
import ghidra.program.model.address.AddressSpace
import ghidra.program.model.address.GenericAddress
import pcodefiles.model.Group.DataOrg
import pcodefiles.model.GroupType
import java.util.HashMap

class Group {
    var id: String
        protected set
    var name: String? = null
    var address: Address? = null

    enum class DataOrg {
        eByte, eLoHi, eHiLo, eLoHiLoHi, eHiloHilo, eFloatLoHi, eFloatHiLo
    }

    var dataOrg: DataOrg? = null
    var groupType: GroupType? = null
    var x = 0
    var y = 0
    var folderName: String? = null

    constructor(id: String) {
        this.id = id
    }

    constructor(id: String, name: String?, address: Address?, dataOrg: String, type: Int, length: Int) {
        this.id = id
        this.name = name
        this.address = address
        y = length
        groupType = intToType(type)
        this.dataOrg = stringToDataOrg(dataOrg)
    }

    private fun intToType(type: Int): GroupType {
        when (type) {
            0 -> return GroupType.GROUP_TYPE_LIST
            1 -> return GroupType.GROUP_TYPE_MAP_2D
            2 -> return GroupType.GROUP_TYPE_MAP_3D
        }
        return GroupType.GROUP_TYPE_LIST
    }

    private fun stringToDataOrg(dataOrg: String): DataOrg {
        when (dataOrg) {
            "eByte" -> return DataOrg.eByte
            "eLoHi" -> return DataOrg.eLoHi
            "eHiLo" -> return DataOrg.eHiLo
            "eLoHiLoHi" -> return DataOrg.eLoHiLoHi
            "eHiLoHilo" -> return DataOrg.eHiloHilo
            "eFloatLoHi" -> return DataOrg.eFloatLoHi
            "eFloatHiLo" -> return DataOrg.eFloatHiLo
        }
        return DataOrg.eByte
    }

    val sizes: Map<String, Int>
        get() {
            val sizes = HashMap<String, Int>()
            sizes["x"] = x
            sizes["y"] = y
            return sizes
        }

    fun setSizes(x: Int, y: Int) {
        this.x = x
        this.y = y
    }

    fun setDataOrg(dataOrg: String) {
        this.dataOrg = stringToDataOrg(dataOrg)
    }

    fun setGroupType(type: Int) {
        groupType = intToType(type)
    }

    val dataTypeSize: Int
        get() {
            when (dataOrg) {
                DataOrg.eByte -> return 1
                DataOrg.eHiLo, DataOrg.eLoHi -> return 2
                DataOrg.eLoHiLoHi, DataOrg.eHiloHilo -> return 4
                DataOrg.eFloatHiLo, DataOrg.eFloatLoHi -> return 8
            }
            return 1
        }

    fun dump(): Map<String, Any?> {
        return mapOf<String, Any?>(
            "id" to id,
            "name" to name,
            "dataOrg" to dataOrg.toString(),
            "address" to address!!.offset.toString(),
            "folderName" to folderName,
            "groupType" to groupType!!.ordinal
        )
    }

    companion object {
        @JvmStatic
        fun unserialize(data: Array<String>): Group {
            return Group(
                data[0],
                data[1],
                null,
                data[3], data[2].toInt(), data[5].toInt()
            )
        }

        @JvmStatic
        fun fromMap(map: Map<String, Any?>): Group {
            return Group(
                map["id"] as String,
                map["name"] as String,
                AddressSpace.EXTERNAL_SPACE.getAddress((map["address"] as String).toLong()),
                map["dataOrg"] as String,
                (map["groupType"] as Double).toInt(),
                0
            )
        }
    }
}